diff --git a/src/modules/meson.build b/src/modules/meson.build
index cdf372c..a25664b 100644
--- a/src/modules/meson.build
+++ b/src/modules/meson.build
@@ -61,6 +61,11 @@ all_modules = [
   [ 'module-virtual-sink', 'module-virtual-sink.c' ],
   [ 'module-virtual-source', 'module-virtual-source.c' ],
   [ 'module-volume-restore', 'module-volume-restore.c' ],
+  
+  # QUBES OS MODULE HERE
+  [ 'module-vchan-sink-la', 'module-vchan-sink.c', ['module-vchan-sink-symdef.h','qubes-vchan-sink.h'], ['-I/usr/include/vchan-xen', '-Dfoomodulepipecustomsymdeffoo'], [], [] ],
+  
+
 ]
 
 if host_machine.system() == 'windows'
@@ -310,7 +315,7 @@ foreach m : all_modules
     install_rpath : rpath_dirs,
     install_dir : modlibexecdir,
     dependencies : [thread_dep, libpulse_dep, libpulsecommon_dep, libpulsecore_dep, libintl_dep, platform_dep, platform_socket_dep] + extra_deps,
-    link_args : [nodelete_link_args, '-Wl,--no-undefined' ],
+    link_args : [nodelete_link_args, '-Wl,--no-undefined', '-lvchan-xen' ],
     link_with : extra_libs,
     name_prefix : '',
     implicit_include_directories : false)
